File: minimum-height-trees.py
class Solution:
    # remove leaves and update leaves from connected branch
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n <= 1:
            return [0]

        graph = collections.defaultdict(list)
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)

        leaves = []
        for node in range(n):
            if len(graph[node]) == 1:
                leaves.append(node)
        
        while n > 2:
            n -= len(leaves)
            new_leaves = []
            
            for leaf in leaves:
                branch = graph[leaf].pop()
                graph[branch].remove(leaf)
                
                if len(graph[branch]) == 1:
                    new_leaves.append(branch)
            
            leaves = new_leaves

        return leaves

    # DFS, iteration, Time Limit Exceeded
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        roots = []
        min_height = n + 1
        graph = collections.defaultdict(list)
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
            
        for root in range(n):
            stack = [(root, 1)]
            visited = collections.defaultdict(int)
            height = 0

            while stack:
                node, h = stack.pop()
                height = max(height, h)
                if not visited[node]:
                    visited[node] = 1
                    for connected in graph[node]:
                        if not visited[connected]:
                            stack.append((connected, h + 1))

            if height == min_height:
                roots.append(root)
            elif height < min_height:
                roots = [root]
                min_height = height
        
        return roots

File: minimum-height-trees.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if(n == 1)
            return Arrays.asList(0);
        
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for(int[] edge: edges) {
            List<Integer> list0;
            List<Integer> list1;

            if(graph.containsKey(edge[0])) {
                list0 = graph.get(edge[0]);
            } else {
                list0 = new ArrayList<>();
            }
            if(graph.containsKey(edge[1])) {
                list1 = graph.get(edge[1]);
            } else {
                list1 = new ArrayList<>();
            }
            
            list0.add(edge[1]);
            list1.add(edge[0]);
            graph.put(edge[0], list0);
            graph.put(edge[1], list1);
        }
        
        List<Integer> leaves = new ArrayList<>();
        for(int i=0; i<n; i++) {
            if(graph.get(i).size() == 1)
                leaves.add(i);
        }

        while(n > 2) {
            n -= leaves.size();
            List<Integer> new_leaves = new ArrayList<>();
            for(int leaf: leaves) {
                int neighbor = graph.get(leaf).get(0);
                graph.get(leaf).remove(0);
                graph.get(neighbor).remove(Integer.valueOf(leaf));

                if(graph.get(neighbor).size() == 1)
                    new_leaves.add(neighbor);
            }

            leaves = new_leaves;
        }

        return leaves;
    }
}